### PR TITLE
Williamhe/1562986 add vso bullseye image

### PR DESCRIPTION
**Pull request by william-msft**
- [x] The purpose of this PR is explained in this message or in an issue. If an issue please include a reference as 

**Goal:**
Add lzma module for python 3.10

**Problem:**
lzma is not supported for the existing vso (githubrunners-buildpackdeps-focal based) image. We know this because lzma comes prebuilt with python versions but not with 3.10. Additionally when we manually add lzma we get kernel related errors.

**Solution:**
Introduce a new vso image that is bullseye based. Bullseye is a newer image with longer support and Oryx is upgrading other images to bullseye too.
This comes with contingencies like removing python 2 (EOL) and the other 2 ruby gems (ruby-debug-ide and debase). Bullseye is a newer image with longer support. This means introducing a new image in Oryx for Codespaces.
https://devdiv.visualstudio.com/DevDiv/_boards/board/t/Oryx/Stories/?workitem=1562986
- [x] Tests are included and/or updated for code changes.
- [x] Proper license headers are included in each file.

### CODE REVIEW

File: build/__variables.sh [hunk 1]
```diff
@@ -75,6 +76,7 @@ declare -r ACR_BUILD_FULL_IMAGE_NAME="$ACR_BUILD_IMAGES_REPO:full"
 declare -r ACR_BUILD_LTS_VERSIONS_IMAGE_NAME="$ACR_BUILD_IMAGES_REPO:lts-versions"
 declare -r ACR_BUILD_GITHUB_ACTIONS_IMAGE_NAME="$ACR_BUILD_IMAGES_REPO:github-actions"
 declare -r ACR_BUILD_VSO_FOCAL_IMAGE_NAME="$ACR_BUILD_IMAGES_REPO:vso-ubuntu-focal"
+declare -r ACR_BUILD_VSO_BULLSEYE_IMAGE_NAME="$ACR_BUILD_IMAGES_REPO:vso-bullseye"
```
Comment (pauld-msft) [inline build/__variables.sh:79]: We now include both OS type and OS version in the image tags. Could you update the tag to be `vso-debian-bullseye`?

File: build/buildBuildImages.sh [hunk 2]
```diff
@@ -514,6 +556,8 @@ elif [ "$imageTypeToBuild" == "full" ]; then
 	buildFullImage "buster"
 elif [ "$imageTypeToBuild" == "vso-focal" ]; then
 	buildVsoFocalImage
```
Comment (pauld-msft) [inline build/buildBuildImages.sh:558]: Looks like the `buildVsoImage` method you made can support a debian bullseye OS version or a ubuntu focal OS version. Should we remove the `buildVsoFocalImage` method and switch this to the `buildVsoImage` you made?

File: images/installPlatform.sh [hunk 3]
```diff
@@ -43,17 +43,14 @@ VERSION="$2"
 
 debianFlavor=$DEBIAN_FLAVOR
 fileName="$PLATFORM_NAME-$VERSION.tar.gz"
-
-sdkStorageAccountUrl="$ORYX_SDK_STORAGE_BASE_URL"
```
Comment (pauld-msft) [inline images/installPlatform.sh:47]: We need this logic so that we can control where the pre-baked binaries come from (dev during testing, prod during actual image release)

File: tests/Oryx.Tests.Common/ImageTestHelper.cs [hunk 4]
```diff
@@ -319,9 +333,10 @@ public string GetGitHubActionsAsBaseWithEnvBuildImage(string debianFlavor = null
             }
         }
 
-        public string GetVsoBuildImage(string debianFlavor=null)
+        public string GetVsoBuildImage(string debianFlavor = null)
         {
-            return $"{_repoPrefix}/{_buildRepository}:{_vsoUbuntu}{_tagSuffix}";
+            string _vsoImage = !string.IsNullOrEmpty(debianFlavor) && debianFlavor == "bullseye" ? _vsoBullseye : _vsoUbuntu;
```
Comment (pauld-msft) [inline tests/Oryx.Tests.Common/ImageTestHelper.cs:338]: Can we make this consistent with the other test methods? They have the image test helper constants (e.g. ImageTestHelperConstants.VsoBullseye) passed in as the `debianFlavor` parameter, instead of the actual debian flavor. This naming can probably be updated as its confusing. 

They then check to see if what is passed in matches the constant e.g.
`string.Equals(debianFlavor.ToLower(), _azureFunctionsJamStackBullseye)` which in this case would be `_vsoBullseye` or `_vsoUbuntu`

All of this allows us to only have to update the `ImageTestHelperConstants` value when an image tag changes, instead of find every place that `GetVsoBuildImage` or `GetBuildImage` is used and seeing if it also needs to be changed

File: images/installPlatform.sh (hunk 3 above)
Comment (pauld-msft) [reply images/installPlatform.sh:47]: Looks like you already reverted this change. Thanks!

File: images/build/Dockerfiles/vso.bullseye.Dockerfile [hunk 5]
```diff
@@ -0,0 +1,325 @@
+FROM mcr.microsoft.com/mirror/docker/library/buildpack-deps:bullseye AS main
```
Comment (cormacpayne) [inline images/build/Dockerfiles/vso.bullseye.Dockerfile:1]: Any reason we don't base this off of our intermediate `oryx/githubrunners-buildpackdeps-bullseye` image?

File: images/build/Dockerfiles/vso.bullseye.Dockerfile [hunk 6]
```diff
@@ -0,0 +1,325 @@
...
+RUN LANG="C.UTF-8" \
+    && apt-get update \
+    && apt-get upgrade -y \
+    && DEBIAN_FRONTEND=noninteractive apt-get install -y --no-install-recommends \
+        git \
+        make \
+        unzip \
+        # The tools in this package are used when installing packages for Python
+        build-essential \
+        # swig3.0 \
```
Comment (cormacpayne) [inline images/build/Dockerfiles/vso.bullseye.Dockerfile:15]: There are a couple of lines that are now commented-out in this file (lines 15, 49, 53-55, 159, 280-285, 296) -- are we able to just delete them?

File: images/build/Dockerfiles/vso.bullseye.Dockerfile [hunk 7]
```diff
@@ -0,0 +1,325 @@
...
+# Even though this adds a new docker layer we are doing this 
+# because we want to avoid duplication (which is always error-prone)
+
+RUN set -ex \
+    && tmpDir="/opt/tmp" \
+    && imagesDir="$tmpDir/images" \
+    && buildDir="$tmpDir/build" \
+    # https://github.com/docker-library/python/issues/147
+    && PYTHONIOENCODING="UTF-8" \    
+    # new next line
```
Comment (cormacpayne) [inline images/build/Dockerfiles/vso.bullseye.Dockerfile:90]: I see this comment a few times in this file now -- are we able to remove it or is it needed for documentation purposes? I'm in the boat that if we want to keep a comment on this line, it should explain what's going on with the `add-apt-repository` command called below it.

File: images/build/Dockerfiles/vso.bullseye.Dockerfile [hunk 8]
```diff
@@ -0,0 +1,325 @@
...
+    && add-apt-repository --remove ppa:xapienz/curl34 \
+    && apt-get update \
+    && apt-get upgrade -y \
+    # It's not clear whether these are needed at runtime...
+    && apt-get install -y --no-install-recommends \
+        tk-dev \
+        uuid-dev \
+        curl \
```
Comment (cormacpayne) [inline images/build/Dockerfiles/vso.bullseye.Dockerfile:98]: nit: is `curl` not included in the base `buildpack-deps` image?

File: images/build/Dockerfiles/vso.bullseye.Dockerfile [hunk 9]
```diff
@@ -0,0 +1,325 @@
...
+    && ln -s $YARN_MINOR_VERSION $YARN_MAJOR_VERSION \
+    # Install Python SDKs
+    # Upgrade system python
+    && mkdir -p /home/codespace/.python \
+    && pip install --upgrade cython \
+    && pip3 install --upgrade cython \
+    && . $buildDir/__pythonVersions.sh \
+    && $imagesDir/installPlatform.sh python $PYTHON39_VERSION \
+    # && $imagesDir/installPlatform.sh python $PYTHON310_VERSION \
+    # build python directly from source
```
Comment (cormacpayne) [inline images/build/Dockerfiles/vso.bullseye.Dockerfile:160]: nit: just curious why we would want to build Python 3.10 from source here rather than use the `installPlatform.sh` script to do this

File: images/build/Dockerfiles/vso.bullseye.Dockerfile [hunk 10]
```diff
@@ -0,0 +1,325 @@
...
+# Now adding remaining of VSO platform features
+RUN buildDir="/opt/tmp/build" \
+    && imagesDir="/opt/tmp/images" \
+    # new next line
+    && add-apt-repository --remove ppa:xapienz/curl34 \
+    && apt-get update \
+    && apt-get install -y --no-install-recommends \
+        apt-transport-https \
+        nano \
+        sed \
```
Comment (cormacpayne) [inline images/build/Dockerfiles/vso.bullseye.Dockerfile:232]: nit: are the `sed` commands you added needed?

File: images/build/Dockerfiles/vso.bullseye.Dockerfile [hunk 11]
```diff
@@ -0,0 +1,325 @@
...
+    && ls . \
+    && php go-pear.phar \
+    #&& mkdir -p pear/tmp/pear/install \
+    #&& wget https://pear.php.net/go-pear.phar \
+    #&& export PHP_PEAR_PHP_BIN=/usr/local/php74/bin/php \
+    #&& export PATH=${HOME}/pear/bin:/usr/local/php74/bin:${PATH} \
+    #&& ls /usr/local/php74/bin/ \
+    #&& /opt/php/8.1.6/bin/php go-pear.phar \
+    && pecl version \
+    && which pecl \
```
Comment (cormacpayne) [inline images/build/Dockerfiles/vso.bullseye.Dockerfile:287]: nit: leftover debugging commands here?

File: images/build/php/prereqs/installPrereqs.sh [hunk 12]
```diff
@@ -22,9 +22,10 @@ PHPIZE_DEPS="autoconf dpkg-dev file g++ gcc libc-dev make pkg-config re2c"
 # libcurl3 and libcurl4 both needs to be supported in ubuntu focal for php
 # https://github.com/xapienz/curl-debian-scripts
 add-apt-repository ppa:xapienz/curl34 -y \
-&& apt-get update \
+apt-get update \
```
Comment (cormacpayne) [inline images/build/php/prereqs/installPrereqs.sh:25]: nit: this works without the `&&` here?

File: images/build/php/prereqs/installPrereqs.sh [hunk 13]
```diff
@@ -22,9 +22,10 @@ PHPIZE_DEPS="autoconf dpkg-dev file g++ gcc libc-dev make pkg-config re2c"
...
 && apt-get upgrade -y \
 && apt-get install -y \
+        sed \
```
Comment (cormacpayne) [inline images/build/php/prereqs/installPrereqs.sh:28]: nit: similar comment as before -- is `sed` (and `grep`) needed?

File: images/build/Dockerfiles/vso.bullseye.Dockerfile (hunk 5 above)
Comment (william-msft) [reply images/build/Dockerfiles/vso.bullseye.Dockerfile:1]: Yeah since this isn't in our approved images: https://eng.ms/docs/more/containers-secure-supply-chain/approved-images

File: images/build/Dockerfiles/vso.bullseye.Dockerfile (hunk 6 above)
Comment (william-msft) [reply images/build/Dockerfiles/vso.bullseye.Dockerfile:15]: Thanks Cormac, I've moved to the PR to draft while I cleanup a little more 😃

File: images/build/Dockerfiles/vso.bullseye.Dockerfile (hunk 8 above)
Comment (william-msft) [reply images/build/Dockerfiles/vso.bullseye.Dockerfile:98]: I was testing with `mcr.microsoft.com/mirror/docker/library/debian:bullseye` which didn't have `curl`. This will be part of my cleanup

File: images/build/Dockerfiles/vso.bullseye.Dockerfile (hunk 7 above)
Comment (william-msft) [reply images/build/Dockerfiles/vso.bullseye.Dockerfile:84]: Thanks for pointing this out, this line should no longer be present.

File: images/build/Dockerfiles/vso.bullseye.Dockerfile (hunk 11 above)
Comment (william-msft) [reply images/build/Dockerfiles/vso.bullseye.Dockerfile:271]: Correct, this has been removed.

File: build/__variables.sh (hunk 1 above)
Comment (william-msft) [reply build/__variables.sh]: Yes, update has been made.

File: build/buildBuildImages.sh (hunk 2 above)
Comment (william-msft) [reply build/buildBuildImages.sh]: Thank Paul, update has been made.

File: tests/Oryx.Tests.Common/ImageTestHelper.cs (hunk 4 above)
Comment (william-msft) [reply tests/Oryx.Tests.Common/ImageTestHelper.cs]: Agreed, I've updated this method.

File: images/build/Dockerfiles/vso.bullseye.Dockerfile (hunk 5 above)
Comment (cormacpayne) [reply images/build/Dockerfiles/vso.bullseye.Dockerfile:1]: That shouldn't affect us since the intermediate images that we build locally are tagged with the prefix `oryxdevmcr.azurecr.io/private/oryx`, so this Dockerfile scan doesn't flag these images because it thinks they're later pushed to our ACR instance, but we just use that naming convention strictly to bypass the check.

File: images/build/Dockerfiles/vso.bullseye.Dockerfile (hunk 10 above)
Comment (william-msft) [reply images/build/Dockerfiles/vso.bullseye.Dockerfile]: Confirmed no longer needed

File: images/build/php/prereqs/installPrereqs.sh (hunk 13 above)
Comment (william-msft) [reply images/build/php/prereqs/installPrereqs.sh:28]: Confirmed no longer needed

File: images/build/php/prereqs/installPrereqs.sh (hunk 12 above)
Comment (william-msft) [reply images/build/php/prereqs/installPrereqs.sh:25]: This actually only works without the `&&` in the front which is strange.

File: images/build/Dockerfiles/vso.bullseye.Dockerfile (hunk 9 above)
Comment (william-msft) [reply images/build/Dockerfiles/vso.bullseye.Dockerfile]: Originally using `installPlatform.sh` for python 3.10 was failing, but I figured out why and have updated this. 

The reason this was failing was because the Debain flavor was still set as focal, which caused focal SDK to be downloaded, which is incompatible with this bullseye image.

File: images/build/Dockerfiles/vso.bullseye.Dockerfile (hunk 8 above)
Comment (william-msft) [reply images/build/Dockerfiles/vso.bullseye.Dockerfile]: Confirmed this is no longer needed.

File: images/build/Dockerfiles/vso.bullseye.Dockerfile (hunk 5 above)
Comment (william-msft) [reply images/build/Dockerfiles/vso.bullseye.Dockerfile:1]: I ran into ` php: error while loading shared libraries: libxml2.so.2: cannot open shared object file: No such file or directory` which is similar to what I saw with `mcr.microsoft.com/mirror/docker/library/debian:bullseye` . Checking to see if I can make it work with this image

File: images/build/php/prereqs/installPrereqs.sh (hunk 13 above)
Comment (cormacpayne) [inline images/build/php/prereqs/installPrereqs.sh:28]: nit: looks like `sed` and `grep` are still being installed with this script -- you mentioned we at least don't need `grep`, so can we revert these changes?

**Review by cormacpayne — DISMISSED** [review]
One minor comment, otherwise :shipit:

File: images/build/php/prereqs/installPrereqs.sh (hunk 13 above)
Comment (william-msft) [reply images/build/php/prereqs/installPrereqs.sh:28]: Thanks for catching this! Its been removed now.

File: platforms/php/versions/bullseye/versionsToBuild.txt [hunk 14]
```diff
@@ -9,6 +9,7 @@
 7.4.28, 9cc3b6f6217b60582f78566b3814532c4b71d517876c25013ae51811e65d8fce, 42670A7FE4D0441C8E4632349E4FDC074A4EF02D 5A52880781F755608BF815FC910DEB46F53EA312,
 7.4.30, ea72a34f32c67e79ac2da7dfe96177f3c451c3eefae5810ba13312ed398ba70d, 42670A7FE4D0441C8E4632349E4FDC074A4EF02D 5A52880781F755608BF815FC910DEB46F53EA312,
 
+8.0.22, 130937c0fa3050cd33d6c415402f6ccbf0682ae83eb8d39c91164224ddfe57f1, 1729F83938DA44E27BA0F4D3DBDB397470D12172 BFDDD28642824F8118EF77909B67A5C12229118F,
```
Comment (william-msft) [inline platforms/php/versions/bullseye/versionsToBuild.txt:12]: We need to add this so the `vso-debian-bullseye` image can download this SDK

File: src/startupscriptgenerator/src/common/Gopkg.toml [hunk 15]
```diff
@@ -41,6 +41,10 @@
   name = "github.com/spf13/afero"
   version = "=1.8.2"
 
+[[override]]
+  name = "golang.org/x/sys"
```
Comment (daniv-msft) [inline src/startupscriptgenerator/src/common/Gopkg.toml:45]: Isn't this coming from Paul's latest PR? Do you need to rebase, or are you adding this to other files?

Comment (william-msft) [reply src/startupscriptgenerator/src/common/Gopkg.toml:45]: I merge Paul's branch `bugfix/lock-sys-unix-package` with my branch since I also need his changes and since they're not in `main` yet

**Review by daniv-msft — APPROVED** [review]
